backend/de.metas.swat/de.metas.swat.base/src/main/java/org/adempiere/invoice/process/C_Invoice_GenerateCorrectionInvoice.java

```diff
@@ -0,0 +1,153 @@
+
+import static org.compiere.model.X_C_DocType.DOCBASETYPE_ARInvoice;
+
+/**
+ * TODO
+ */
+public class C_Invoice_GenerateCorrectionInvoice extends JavaProcess implements IProcessPrecondition
+{
+	private final IInvoiceDAO invoiceDAO = Services.get(IInvoiceDAO.class);
+	private final IInvoiceBL invoiceBL = Services.get(IInvoiceBL.class);
+	private final IDocTypeDAO docTypeDAO = Services.get(IDocTypeDAO.class);
+	private final ICountryDAO countryDAO = Services.get(ICountryDAO.class);
+	private final INotificationBL notificationBL = Services.get(INotificationBL.class);
+	private final IUserBL userBL = Services.get(IUserBL.class);
+
+	private static final DocBaseAndSubType SALES_INVOICE = DocBaseAndSubType.of(DOCBASETYPE_ARInvoice);
+	private static final DocBaseAndSubType CORRECTION_INVOICE = DocBaseAndSubType.of(X_C_DocType.DOCBASETYPE_ARInvoice, X_C_DocType.DOCSUBTYPE_CorrectionInvoice);
+	private static final int C_DocType_ID = -1;
+	private static final AdMessageKey MSG_Event_DocumentGenerated = AdMessageKey.of("Event_DocumentGenerated");
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNull IProcessPreconditionsContext context)
+	{
+		if (!context.isSingleSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection().toInternal();
+		}
+
+		final InvoiceId invoiceId = InvoiceId.ofRepoId(context.getSingleSelectedRecordId());
+		final I_C_Invoice invoiceRecord = invoiceDAO.getByIdInTrx(invoiceId);
+		final DocTypeId docTypeId = DocTypeId.ofRepoId(invoiceRecord.getC_DocType_ID());
+		final DocBaseAndSubType docBaseAndSubType = docTypeDAO.getDocBaseAndSubTypeById(docTypeId);
+
+		if(!(docBaseAndSubType.equals(SALES_INVOICE) || invoiceBL.isAdjustmentCharge(invoiceRecord)))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not an basic Sales Invoice or Adjustment Charge");
+		}
+
+		final DocStatus docStatus = DocStatus.ofCode(invoiceRecord.getDocStatus());
+		if(!docStatus.isCompletedOrClosed())
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not completed or closed");
+		}
+
+		if(!countryDAO.isEnforceCorrectionInvoice(invoiceBL.getBillToCountryId(invoiceId)))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not isEnforceCorrectionInvoice");
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final de.metas.adempiere.model.I_C_Invoice invoice = InterfaceWrapperHelper.create(getCtx(), getRecord_ID(), de.metas.adempiere.model.I_C_Invoice.class, get_TrxName());
+
+		final InvoiceCreditContext creditCtx = InvoiceCreditContext.builder()
+				.docTypeId(null)
+				.completeAndAllocate(true)
+				.referenceOriginalOrder(true)
+				.referenceInvoice(true)
+				.creditedInvoiceReinvoicable(false)
+				.fixedInvoice(true)
+				.build();
+
+		final I_C_Invoice creditMemo = invoiceBL.creditInvoice(invoice, creditCtx);
```

Comment:
maybe u could just provide the invoiceId instead of loaded invoice.
I saw that u are not using it here anyways.

More, as a result you could consider returning a CreditInvoiceResult { InvoiceId originalInvoiceId; I_C_Invoice creditInvoice; }

Then u you could use the originalInvoiceId when creating the AdjustmentChargeCreateRequest.